### ISSUE TITLE
🐛Ad for following adId already exists: 0

### ISSUE THREAD

**Issue by Tetracyl** [googleads/googleads-mobile-flutter#47]
## Bug report

**Describe the bug**
This error appears if I set the adUnitId to literally anything. I've tried setting the property to 'hi' and a bunch of other stuff but it somehow still appears.  And if I set it to a valid ad id it continues to appear and the ad never shows.

```E/MethodChannel#plugins.flutter.io/google_mobile_ads( 6263): Failed to handle method call
E/MethodChannel#plugins.flutter.io/google_mobile_ads( 6263): java.lang.IllegalArgumentException: Ad for following adId already exists: 0
E/MethodChannel#plugins.flutter.io/google_mobile_ads( 6263):    at io.flutter.plugins.googlemobileads.AdInstanceManager.trackAd(AdInstanceManager.java:66)
E/MethodChannel#plugins.flutter.io/google_mobile_ads( 6263):    at io.flutter.plugins.googlemobileads.GoogleMobileAdsPlugin.onMethodCall(GoogleMobileAdsPlugin.java:266)
E/MethodChannel#plugins.flutter.io/google_mobile_ads( 6263):    at io.flutter.plugin.common.MethodChannel$IncomingMethodCallHandler.onMessage(MethodChannel.java:233)
E/MethodChannel#plugins.flutter.io/google_mobile_ads( 6263):    at io.flutter.embedding.engine.dart.DartMessenger.handleMessageFromDart(DartMessenger.java:85)
E/MethodChannel#plugins.flutter.io/google_mobile_ads( 6263):    at io.flutter.embedding.engine.FlutterJNI.handlePlatformMessage(FlutterJNI.java:692)
E/MethodChannel#plugins.flutter.io/google_mobile_ads( 6263):    at android.os.MessageQueue.nativePollOnce(Native Method)
E/MethodChannel#plugins.flutter.io/google_mobile_ads( 6263):    at android.os.MessageQueue.next(MessageQueue.java:335)
E/MethodChannel#plugins.flutter.io/google_mobile_ads( 6263):    at android.os.Looper.loop(Looper.java:183)
E/MethodChannel#plugins.flutter.io/google_mobile_ads( 6263):    at android.app.ActivityThread.main(ActivityThread.java:7656)
E/MethodChannel#plugins.flutter.io/google_mobile_ads( 6263):    at java.lang.reflect.Method.invoke(Native Method)
E/MethodChannel#plugins.flutter.io/google_mobile_ads( 6263):    at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:592)
E/MethodChannel#plugins.flutter.io/google_mobile_ads( 6263):    at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:947)
E/flutter ( 6263): [ERROR:flutter/lib/ui/ui_dart_state.cc(177)] Unhandled Exception: PlatformException(error, Ad for following adId already exists: 0, null, java.lang.IllegalArgumentException: Ad for following adId already exists: 0
E/flutter ( 6263):      at io.flutter.plugins.googlemobileads.AdInstanceManager.trackAd(AdInstanceManager.java:66)
E/flutter ( 6263):      at io.flutter.plugins.googlemobileads.GoogleMobileAdsPlugin.onMethodCall(GoogleMobileAdsPlugin.java:266)
E/flutter ( 6263):      at io.flutter.plugin.common.MethodChannel$IncomingMethodCallHandler.onMessage(MethodChannel.java:233)
E/flutter ( 6263):      at io.flutter.embedding.engine.dart.DartMessenger.handleMessageFromDart(DartMessenger.java:85)
E/flutter ( 6263):      at io.flutter.embedding.engine.FlutterJNI.handlePlatformMessage(FlutterJNI.java:692)
E/flutter ( 6263):      at android.os.MessageQueue.nativePollOnce(Native Method)
E/flutter ( 6263):      at android.os.MessageQueue.next(MessageQueue.java:335)
E/flutter ( 6263):      at android.os.Looper.loop(Looper.java:183)
E/flutter ( 6263):      at android.app.ActivityThread.main(ActivityThread.java:7656)
E/flutter ( 6263):      at java.lang.reflect.Method.invoke(Native Method)
E/flutter ( 6263):      at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:592)
E/flutter ( 6263):      at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:947)
E/flutter ( 6263): )
E/flutter ( 6263): #0      StandardMethodCodec.decodeEnvelope (package:flutter/src/services/message_codecs.dart:582:7)
E/flutter ( 6263): #1      MethodChannel._invokeMethod (package:flutter/src/services/platform_channel.dart:159:18)
E/flutter ( 6263): <asynchronous suspension>
E/flutter ( 6263): #2      MethodChannel.invokeMethod (package:flutter/src/services/platform_channel.dart:332:12)
E/flutter ( 6263): #3      AdInstanceManager.loadNativeAd (package:google_mobile_ads/src/ad_instance_manager.dart:172:20)
E/flutter ( 6263): #4      NativeAd.load (package:google_mobile_ads/src/ad_containers.dart:576:27)
E/flutter ( 6263): #5      NativeAdState.initState.<anonymous closure> (package:discord_bots/widgets/global/native_ad_widget.dart:36:65)
E/flutter ( 6263): #6      new Future.delayed.<anonymous closure> (dart:async/future.dart:326:39)
E/flutter ( 6263): #7      _rootRun (dart:async/zone.dart:1182:47)
E/flutter ( 6263): #8      _CustomZone.run (dart:async/zone.dart:1093:19)
E/flutter ( 6263): #9      _CustomZone.runGuarded (dart:async/zone.dart:997:7)
E/flutter ( 6263): #10     _CustomZone.bindCallbackGuarded.<anonymous closure> (dart:async/zone.dart:1037:23)
E/flutter ( 6263): #11     _rootRun (dart:async/zone.dart:1190:13)
E/flutter ( 6263): #12     _CustomZone.run (dart:async/zone.dart:1093:19)
E/flutter ( 6263): #13     _CustomZone.bindCallback.<anonymous closure> (dart:async/zone.dart:1021:23)
E/flutter ( 6263): #14     Timer._createTimer.<anonymous closure> (dart:async-patch/timer_patch.dart:18:15)
E/flutter ( 6263): #15     _Timer._runTimers (dart:isolate-patch/timer_impl.dart:397:19)
E/flutter ( 6263): #16     _Timer._handleMessage (dart:isolate-patch/timer_impl.dart:428:5)
E/flutter ( 6263): #17     _RawReceivePortImpl._handleMessage (dart:isolate-patch/isolate_patch.dart:168:12)
E/flutter ( 6263):
```

### Steps to reproduce

1. Create a native ad widget
2. Add the java files to the src folder and add the layout XML file to the res folder.
3. Run app

---

## Additional context
My code is from the example:
```import 'dart:async';

import 'package:flutter/material.dart';
import 'package:google_mobile_ads/google_mobile_ads.dart';

class NativeAdWidget extends StatefulWidget {
  @override
  State<StatefulWidget> createState() => NativeAdState();
}

class NativeAdState extends State<NativeAdWidget> {
  NativeAd _nativeAd;
  final Completer<NativeAd> nativeAdCompleter = Completer<NativeAd>();

  @override
  void initState() {
    super.initState();
    _nativeAd = NativeAd(
      adUnitId: NativeAd.testAdUnitId,
      request: AdRequest(),
      factoryId: 'listTile',
      listener: AdListener(
        onAdLoaded: (Ad ad) {
          print('$NativeAd loaded.');
          nativeAdCompleter.complete(ad as NativeAd);
        },
        onAdFailedToLoad: (Ad ad, LoadAdError error) {
          print('$NativeAd failedToLoad: $error');
          nativeAdCompleter.completeError(null);
        },
        onAdOpened: (Ad ad) => print('$NativeAd onAdOpened.'),
        onAdClosed: (Ad ad) => print('$NativeAd onAdClosed.'),
        onApplicationExit: (Ad ad) => print('$NativeAd onApplicationExit.'),
      ),
    );
    Future<void>.delayed(Duration(seconds: 1), () => _nativeAd?.load());
  }

  @override
  void dispose() {
    super.dispose();
    _nativeAd?.dispose();
    _nativeAd = null;
  }

  @override
  Widget build(BuildContext context) {
    return FutureBuilder<NativeAd>(
      future: nativeAdCompleter.future,
      builder: (BuildContext context, AsyncSnapshot<NativeAd> snapshot) {
        Widget child;

        switch (snapshot.connectionState) {
          case ConnectionState.none:
          case ConnectionState.waiting:
          case ConnectionState.active:
            child = Container();
            break;
          case ConnectionState.done:
            if (snapshot.hasData) {
              child = AdWidget(ad: _nativeAd);
            } else {
              child = Text('Error loading $NativeAd');
            }
        }

        return Container(
          width: 250,
          height: 350,
          child: child,
          color: Colors.blueGrey,
        );
      },
    );
  }
}
```

---

**Comment by DennisH511** [comment]
Ads are loading and displaying when I run the app on a device or in the simulator but will crash if I press the "Restart" button - not the "Hot Reload" button.

Error message:
````
E/MethodChannel#plugins.flutter.io/google_mobile_ads(14986): Failed to handle method call
E/MethodChannel#plugins.flutter.io/google_mobile_ads(14986): java.lang.IllegalArgumentException: Ad for following adId already exists: 0
E/MethodChannel#plugins.flutter.io/google_mobile_ads(14986): 	at io.flutter.plugins.googlemobileads.AdInstanceManager.trackAd(AdInstanceManager.java:66)
E/MethodChannel#plugins.flutter.io/google_mobile_ads(14986): 	at io.flutter.plugins.googlemobileads.GoogleMobileAdsPlugin.onMethodCall(GoogleMobileAdsPlugin.java:244)
E/MethodChannel#plugins.flutter.io/google_mobile_ads(14986): 	at io.flutter.plugin.common.MethodChannel$IncomingMethodCallHandler.onMessage(MethodChannel.java:233)
E/MethodChannel#plugins.flutter.io/google_mobile_ads(14986): 	at io.flutter.embedding.engine.dart.DartMessenger.handleMessageFromDart(DartMessenger.java:85)
E/MethodChannel#plugins.flutter.io/google_mobile_ads(14986): 	at io.flutter.embedding.engine.FlutterJNI.handlePlatformMessage(FlutterJNI.java:692)
E/MethodChannel#plugins.flutter.io/google_mobile_ads(14986): 	at android.os.MessageQueue.nativePollOnce(Native Method)
E/MethodChannel#plugins.flutter.io/google_mobile_ads(14986): 	at android.os.MessageQueue.next(MessageQueue.java:336)
E/MethodChannel#plugins.flutter.io/google_mobile_ads(14986): 	at android.os.Looper.loop(Looper.java:174)
E/MethodChannel#plugins.flutter.io/google_mobile_ads(14986): 	at android.app.ActivityThread.main(ActivityThread.java:7356)
E/MethodChannel#plugins.flutter.io/google_mobile_ads(14986): 	at java.lang.reflect.Method.invoke(Native Method)
E/MethodChannel#plugins.flutter.io/google_mobile_ads(14986): 	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:492)
E/MethodChannel#plugins.flutter.io/google_mobile_ads(14986): 	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:930)
````

The app crashes at the following line in the Widget: _bannerAd?.load();

Banner Widget code:
````
class TestBanner extends StatefulWidget {
  TestBanner({Key key}) : super(key: key);

  @override
  _TestBannerState createState() => _TestBannerState();
}

class _TestBannerState extends State<TestBanner> {
  BannerAd _bannerAd;

  @override
  void initState() {
    super.initState();
    _bannerAd = BannerAd(
      adUnitId: BannerAd.testAdUnitId,
      request: AdRequest(),
      size: AdSize.banner,
      listener: AdListener(
        onAdLoaded: (Ad ad) {
          print('$BannerAd loaded.');
        },
        onAdFailedToLoad: (Ad ad, LoadAdError error) {
          print('$BannerAd failedToLoad: $error');
        },
        onAdOpened: (Ad ad) => print('$BannerAd onAdOpened.'),
        onAdClosed: (Ad ad) => print('$BannerAd onAdClosed.'),
        onApplicationExit: (Ad ad) => print('$BannerAd onApplicationExit.'),
      ),
    );
    _bannerAd?.load();
  }

  @override
  Widget build(BuildContext context) {
    final AdWidget adWidget = AdWidget(ad: _bannerAd);
    return Container(
      alignment: Alignment.center,
      child: adWidget,
      width: _bannerAd.size.width.toDouble(),
      height: _bannerAd.size.height.toDouble(),
    );
  }

  @override
  void dispose() {
    super.dispose();
    _bannerAd?.dispose();
    _bannerAd = null;
  }
}
````

**Comment by MaximeDelhorme** [comment]
Also getting the error.

**Comment by jimmyff** [comment]
@jjliu15 apologies to tag you but is there any chance you or your colleagues could have a look at this issue? I have an app update that needs to go live but is currently blocked by this.

**Comment by jjliu15** [comment]
Is this still happening in the latest version of the plugin? (0.11.0+3)
Also could you share the output of `flutter doctor`?

**Comment by Tetracyl** [comment]
I tried upgrading to the latest version of the plugin and now I'm receiving multiple different errors. When I did `flutter run` I got this error:
```
I/Ads     (24285): Received log message: <Google:HTML> Incorrect native ad response. Click actions were not properly specified
I/Ads     (24285): Ad failed to load : 0
I/flutter (24285): NativeAd failedToLoad: LoadAdError(code: 0, domain: com.google.android.gms.ads, message: Internal error.)
```
After running a hot restart I got the `Ad for following adId already exists: 0` error again.

When I tried running the example app I didn't encounter any errors though.



### Flutter Doctor:
<details><summary>Click To Expand</summary>

```
Doctor summary (to see all details, run flutter doctor -v):
[√] Flutter (Channel stable, 1.22.6, on Microsoft Windows [Version 10.0.19042.804], locale en-US)
[√] Android toolchain - develop for Android devices (Android SDK version 30.0.3)
[!] Android Studio (version 4.1.0)
    X Flutter plugin not installed; this adds Flutter specific functionality.   
    X Dart plugin not installed; this adds Dart specific functionality.
[√] VS Code (version 1.54.1)        
[√] Connected device (1 available)  

! Doctor found issues in 1 category.
```

**Comment by jjliu15** [comment]
@bparrishMines  FYI

The `Ad for following adId already exists: 0` error is due to internal plugin state not getting cleaned up on `hot restart`. From https://github.com/flutter/flutter/issues/7160, there is currently no hook that notifies plugins of hot restart. Seems like the the camera plugin has a [workaround](https://github.com/flutter/flutter/issues/10437#issuecomment-457653187) for this, which I'll look into implementing for the ads plugin. 

Note this is not an error that should affect production apps, but it will hinder development. For now, please reload the app instead of using hot reload during development.

**Comment by jjliu15** [comment]
@Tetracyl Your new error is an internal error coming from the ads sdk, not the plugin. 

Does it happen when using the test ad unit? (`ca-app-pub-3940256099942544/2247696110`)
Also have you followed the setup steps in https://github.com/googleads/googleads-mobile-flutter/blob/master/README.md#platform-specific-setup?

**Comment by Tetracyl** [comment]
When I use the test ad unit I get this error after doing `flutter run`:

```
I/Ads     ( 3197): Ad failed to load : 3
I/flutter ( 3197): NativeAd failedToLoad: LoadAdError(code: 3, domain: com.google.android.gms.ads, message: No ad config.)
```

If I try to hot restart I get the `Ad for following adId already exists: 0` error again. And yes I have followed the setup steps.

**Comment by icaksama** [comment]
Dispose of the ad on the listener `onAdFailedToLoad`. Stop and run the app again.
```
onAdFailedToLoad: (Ad ad, LoadAdError error) {
            print('BannerAd failedToLoad: $error');
            _bannerAd?.dispose();
            _bannerAd = null;
            nativeAdCompleter.completeError(null);
},
```

**Comment by jimmyff** [comment]
After upgrading to the latest package I started getting this error:
```
java.lang.IllegalArgumentException: Service not registered: com.google.android.gms.measurement.internal.zzjl@95dd0fd
```

I added this to my `build.gradle;' which seems to have resolved the issue:
```
    implementation 'com.google.android.gms:play-services-basement:17.5.0'
```

If this is required shouldn't it be included in the installation steps?

**Comment by Tetracyl** [comment]
> 
> 
> Dispose of the ad on the listener `onAdFailedToLoad`. Stop and run the app again.
> 
> ```
> onAdFailedToLoad: (Ad ad, LoadAdError error) {
>             print('BannerAd failedToLoad: $error');
>             _bannerAd?.dispose();
>             _bannerAd = null;
>             nativeAdCompleter.completeError(null);
> },
> ```
> 
> Change `initState` to `didChangeDependencies`:
> 
> ```
> @override
>   void didChangeDependencies() {
>     super.didChangeDependencies();
>     if (widget.isEnable) {
>       _bannerAd = BannerAd(
>         adUnitId: BannerAd.testAdUnitId,
>         size: AdSize(width: (Helper.getWidth(context)).toInt(), height: 330),
>         request: AdRequest(),
>         listener: AdListener(
>           onAdLoaded: (Ad ad) {
>             print('BannerAd loaded.');
>             bannerAdCompleter.complete(ad as BannerAd);
>           },
>           onAdFailedToLoad: (Ad ad, LoadAdError error) {
>             print('BannerAd failedToLoad: $error');
>             _bannerAd?.dispose();
>             _bannerAd = null;
>             bannerAdCompleter.completeError(null);
>           },
>           onAdOpened: (Ad ad) => print('BannerAd onAdOpened.'),
>           onAdClosed: (Ad ad) => print('BannerAd onAdClosed.'),
>           onApplicationExit: (Ad ad) => print('BannerAd onApplicationExit.'),
>         ),
>       )..load();
>     }
>   }
> ```

Ok I tried disposing of the ad and changing initState to didChangeDependencies but now I'm getting a different error:
```
I/Ads     ( 4652): Ad failed to load : 3
I/flutter ( 4652): NativeAd failedToLoad: LoadAdError(code: 3, domain: com.google.android.gms.ads, message: No ad config.)
```

Code:
```
  @override
  void didChangeDependencies() {
    super.didChangeDependencies();
    _nativeAd = NativeAd(
      adUnitId: NativeAd.testAdUnitId,
      request: AdRequest(),
      factoryId: 'listTile',
      listener: AdListener(
        onAdLoaded: (Ad ad) {
          print('$NativeAd loaded.');
          nativeAdCompleter.complete(ad as NativeAd);
        },
        onAdFailedToLoad: (Ad ad, LoadAdError error) {
          print('$NativeAd failedToLoad: $error');
          _nativeAd?.dispose();
          _nativeAd = null;
          nativeAdCompleter.completeError(null);
        },
        onAdOpened: (Ad ad) => print('$NativeAd onAdOpened.'),
        onAdClosed: (Ad ad) => print('$NativeAd onAdClosed.'),
        onApplicationExit: (Ad ad) => print('$NativeAd onApplicationExit.'),
      ),
    );
    Future<void>.delayed(Duration(seconds: 1), () => _nativeAd?.load());
  }
```

Will there be a fix soon or should I try using another ad package since I too have an app update that needs to go live and these issues are preventing me from pushing the update.

**Comment by daybson** [comment]
Same here. Glat I'm not alone.

**Comment by icaksama** [comment]
@Tetracyl Native ad has the issue, I have no clue. Try to use this version `google_mobile_ads: ^0.11.0+2`, it works for others but not me. But banner ad working properly.

To make sure the ad always shows in your app, please load a banner ad as an alternative when the native ad is failed to load.

Create banner and native ad reusable widget.
For example in native ad reusable widget:
```
@override
  Widget build(BuildContext context) {
    super.build(context);
    if (widget.isEnable) {
      return FutureBuilder(
        future: nativeAdCompleter.future,
        builder: (BuildContext context, snapshot) {
          Widget child;

          switch (snapshot.connectionState) {
            case ConnectionState.none:
            case ConnectionState.waiting:
            case ConnectionState.active:
              // CHANGE THIS WITH YOUR BANNER
              child = FixBannerAdmob(
                widget.adUnitId,
                widget.margin,
                isEnable: widget.isEnable,
              );
              break;
            case ConnectionState.done:
              if (snapshot.hasData) {
                child = AdWidget(ad: _nativeAd);
              } else {
                child = Text('Error loading $NativeAd');
              }
          }

          return Container(
            height: 330,
            margin: widget.margin,
            decoration: BoxDecoration(
              color: Colors.blueGrey.withOpacity(0.7),
              image: DecorationImage(
                colorFilter: ColorFilter.mode(
                    Colors.black.withOpacity(0.6), BlendMode.dstATop),
                image: Image.asset('assets/images/placeholder.jpg').image,
                fit: BoxFit.cover,
              ),
            ),
            child: child,
          );
        },
      );
    }
    return Container();
  }
```

**Comment by WieFel** [comment]
Still having same issues with newest version.
Using version `google_mobile_ads: ^0.11.0+3` and Android Admob test unit id.

My app consists of 2 pages. Both display a banner at the bottom. On app start, page 1 is shown.
Each page has its own `BannerAd` object stored in a state property. In `didChangeDependencies()`, the BannerAd is instantiated. On page `dispose()`, the BannerAd is disposed as well.

### On Android emulator using Android 11
- When freshly starting the app -> no ads are displaying at all
```
I/Ads     (16383): Ad failed to load : 0
```
`AdListener.onAdFailedToLoad` outputs:
```
LoadAdError(code: 0, domain: com.google.android.gms.ads, message: Internal error.)
```
### On Pixel 5 using Android 11
1. On initial app start -> test ad is showing correctly on page 1. `onAdLoaded` outputs `ca-app-pub-3940256099942544/6300978111`.
2. However, after switching to page 2, no ad is shown there (although `onAdLoaded` outputs the same as in 1.). When switching back to page 1, no ad is shown anymore either, app gets stuck and error occurs:
```
Unhandled Exception: PlatformException(error, Ad for following adId already exists: 2, null, java.lang.IllegalArgumentException: Ad for following adId already exists: 2
```
3. Hot restart in this case resolves the problem and process starts as in **1.**.

<br/>
<details>
  <summary>Flutter Doctor Output</summary>
  
  Doctor summary (to see all details, run flutter doctor -v):
  [✓] Flutter (Channel stable, 2.0.2, on Linux, locale en_GB.UTF-8)
  [✓] Android toolchain - develop for Android devices (Android SDK version 29.0.3)
  [✓] Chrome - develop for the web
  [✓] Android Studio (version 4.0)
  [✓] Connected device (2 available)
  
  • No issues found!
</details>

**Comment by yosus** [comment]
I am using `google_mobile_ads: ^0.11.0+3`

Testing on a device using `BannerAd.testAdUnitId`. Also see the error if I do a Run (Shift+F10). When it happens, I need to kill the app on the device and reload it. The error doesn't appear if using 'hot reload' (Control + \)


Log :
```
E/MethodChannel#plugins.flutter.io/google_mobile_ads( 1648): Failed to handle method call
E/MethodChannel#plugins.flutter.io/google_mobile_ads( 1648): java.lang.IllegalArgumentException: Ad for following adId already exists: 0
E/MethodChannel#plugins.flutter.io/google_mobile_ads( 1648): 	at io.flutter.plugins.googlemobileads.AdInstanceManager.trackAd(AdInstanceManager.java:66)
E/MethodChannel#plugins.flutter.io/google_mobile_ads( 1648): 	at io.flutter.plugins.googlemobileads.GoogleMobileAdsPlugin.onMethodCall(GoogleMobileAdsPlugin.java:244)
E/MethodChannel#plugins.flutter.io/google_mobile_ads( 1648): 	at io.flutter.plugin.common.MethodChannel$IncomingMethodCallHandler.onMessage(MethodChannel.java:233)
E/MethodChannel#plugins.flutter.io/google_mobile_ads( 1648): 	at io.flutter.embedding.engine.dart.DartMessenger.handleMessageFromDart(DartMessenger.java:85)
E/MethodChannel#plugins.flutter.io/google_mobile_ads( 1648): 	at io.flutter.embedding.engine.FlutterJNI.handlePlatformMessage(FlutterJNI.java:692)
E/MethodChannel#plugins.flutter.io/google_mobile_ads( 1648): 	at android.os.MessageQueue.nativePollOnce(Native Method)
E/MethodChannel#plugins.flutter.io/google_mobile_ads( 1648): 	at android.os.MessageQueue.next(MessageQueue.java:336)
E/MethodChannel#plugins.flutter.io/google_mobile_ads( 1648): 	at android.os.Looper.loop(Looper.java:197)
E/MethodChannel#plugins.flutter.io/google_mobile_ads( 1648): 	at android.app.ActivityThread.main(ActivityThread.java:8167)
E/MethodChannel#plugins.flutter.io/google_mobile_ads( 1648): 	at java.lang.reflect.Method.invoke(Native Method)
E/MethodChannel#plugins.flutter.io/google_mobile_ads( 1648): 	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:496)
E/MethodChannel#plugins.flutter.io/google_mobile_ads( 1648): 	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1100)
I/flutter ( 1648): -------- caught: PlatformException(error, Ad for following adId already exists: 0, null, java.lang.IllegalArgumentException: Ad for following adId already exists: 0
I/flutter ( 1648): 	at io.flutter.plugins.googlemobileads.AdInstanceManager.trackAd(AdInstanceManager.java:66)
I/flutter ( 1648): 	at io.flutter.plugins.googlemobileads.GoogleMobileAdsPlugin.onMethodCall(GoogleMobileAdsPlugin.java:244)
I/flutter ( 1648): 	at io.flutter.plugin.common.MethodChannel$IncomingMethodCallHandler.onMessage(MethodChannel.java:233)
I/flutter ( 1648): 	at io.flutter.embedding.engine.dart.DartMessenger.handleMessageFromDart(DartMessenger.java:85)
I/flutter ( 1648): 	at io.flutter.embedding.engine.FlutterJNI.handlePlatformMessage(FlutterJNI.java:692)
I/flutter ( 1648): 	at android.os.MessageQueue.nativePollOnce(Native Method)
I/flutter ( 1648): 	at android.os.MessageQueue.next(MessageQueue.java:336)
I/flutter ( 1648): 	at android.os.Looper.loop(Looper.java:197)
I/flutter ( 1648): 	at android.app.ActivityThread.main(ActivityThread.java:8167)
I/flutter ( 1648): 	at java.lang.reflect.Method.invoke(Native Method)
I/flutter ( 1648): 	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:496)
I/flutter ( 1648): 	at com.android.in
W/ConnectionTracker( 1648): Exception thrown while unbinding
W/ConnectionTracker( 1648): java.lang.IllegalArgumentException: Service not registered: com.google.android.gms.measurement.internal.zzji@e9874fa
W/ConnectionTracker( 1648): 	at android.app.LoadedApk.forgetServiceDispatcher(LoadedApk.java:1870)
W/ConnectionTracker( 1648): 	at android.app.ContextImpl.unbindService(ContextImpl.java:1848)
W/ConnectionTracker( 1648): 	at android.content.ContextWrapper.unbindService(ContextWrapper.java:755)
W/ConnectionTracker( 1648): 	at com.google.android.gms.common.stats.ConnectionTracker.zza(com.google.android.gms:play-services-basement@@17.3.0:55)
W/ConnectionTracker( 1648): 	at com.google.android.gms.common.stats.ConnectionTracker.unbindService(com.google.android.gms:play-services-basement@@17.3.0:50)
W/ConnectionTracker( 1648): 	at com.google.android.gms.measurement.internal.zzio.zzag(com.google.android.gms:play-services-measurement-impl@@17.6.0:245)
W/ConnectionTracker( 1648): 	at com.google.android.gms.measurement.internal.zzio.zzal(com.google.android.gms:play-services-measurement-impl@@17.6.0:262)
W/ConnectionTracker( 1648): 	at com.google.android.gms.measurement.internal.zzio.zzc(com.google.android.gms:play-services-measurement-impl@@17.6.0:336)
W/ConnectionTracker( 1648): 	at com.google.android.gms.measurement.internal.zzir.zza(com.google.android.gms:play-services-measurement-impl@@17.6.0:2)
W/ConnectionTracker( 1648): 	at com.google.android.gms.measurement.internal.zzai.run(com.google.android.gms:play-services-measurement-impl@@17.6.0:7)
W/ConnectionTracker( 1648): 	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:462)
W/ConnectionTracker( 1648): 	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
W/ConnectionTracker( 1648): 	at com.google.android.gms.measurement.internal.zzfs.run(com.google.android.gms:play-services-measurement-impl@@17.6.0:21)

```

**Comment by WieFel** [comment]
Problem solved for me!
By...
1. updating to `google_mobile_ads: ^0.11.0+4` (this alone doesn't fix anything yet)
2. taking `BannerAdWidget` from the [example](https://github.com/googleads/googleads-mobile-flutter/blob/b32a33b6fd93b4fd0a0073c22196e925c15f2bc0/packages/google_mobile_ads/example/lib/main.dart#L217-L292) project and adapting it to my needs.

https://github.com/googleads/googleads-mobile-flutter/blob/b32a33b6fd93b4fd0a0073c22196e925c15f2bc0/packages/google_mobile_ads/example/lib/main.dart#L217-L292

Apparently the `Completer<BannerAd>` that is used there is making it work for me!

**Comment by jjliu15** [comment]
The crash on hot reload should be resolved in `0.11.0+4`, via https://github.com/googleads/googleads-mobile-flutter/pull/94.

@WieFel If you're having issue getting test ads to work could you open a separate issue?

**Comment by RamsayGit** [comment]
I'm still getting this error on hot reload:
I/flutter (11835): PlatformException(error, Ad for following adId already exists: 0, null, java.lang.IllegalArgumentException: Ad for following adId already exists: 0
I/flutter (11835): 	at io.flutter.plugins.googlemobileads.AdInstanceManager.trackAd(AdInstanceManager.java:79)
I/flutter (11835): 	at io.flutter.plugins.googlemobileads.GoogleMobileAdsPlugin.onMethodCall(GoogleMobileAdsPlugin.java:458)
I/flutter (11835): 	at io.flutter.plugin.common.MethodChannel$IncomingMethodCallHandler.onMessage(MethodChannel.java:258)
I/flutter (11835): 	at io.flutter.embedding.engine.dart.DartMessenger.invokeHandler(DartMessenger.java:295)
I/flutter (11835): 	at io.flutter.embedding.engine.dart.DartMessenger.lambda$dispatchMessageToQueue$0$io-flutter-embedding-engine-dart-DartMessenger(DartMessenger.java:322)
I/flutter (11835): 	at io.flutter.embedding.engine.dart.DartMessenger$$ExternalSyntheticLambda0.run(Unknown Source:12)
I/flutter (11835): 	at android.os.Handler.handleCallback(Handler.java:942)
I/flutter (11835): 	at android.os.Handler.dispatchMessage(Handler.java:99)
I/flutter (11835): 	at android.os.Looper.loopOnce(Looper.java:201)
I/flutter (11835): 	at android.os.Looper.loop(Looper.java:288)
I/flutter (11835): 	at android.app.ActivityThr
I/flutter (11835): #0      StandardMethodCodec.decodeEnvelope (package:flutter/src/services/message_codecs.dart:653:7)
I/flutter (11835): #1      MethodChannel._invokeMethod (package:flutter/src/services/platform_channel.dart:315:18)
I/flutter (11835): <asynchronous suspension>
I/flutter (11835): #2      RewardedAd.load (package:google_mobile_ads/src/ad_containers.dart:1195:5)
I/flutter (11835): <asynchronous suspension>
